btclightclient/partial_merkle_tree.go

```diff
@@ -0,0 +1,271 @@
+package btclightclient
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// We verify a partial merkle tree proof. This logic is used in verifytxoutproof.
+// First one, we must know how merkle tree store in bitcoin. The tree is prefect
+// binary tree. At height `h` this index from 0 to `tree width` at level h.
+// At level(height) 0, this is the list of transaction IDs. At level 1, we compute
+// node value at positon `p` = hash(node at (h - 1, p * 2), node at (h - 1, p * 2 + 1))
+// You can see p * 2 is left node and p * 2 + 1 is right node. We define (h, p)
+// is  merkle node at height h, position p on this level. We can virualize the tree
+// like below:
+//
+// level h                              (h, 0)
+//                                  /              \
+// level h - 1                  (h-1, 0)             (h- 1, 1)
+//                             /       \             /         \
+// level h - 2              (h - 2, 0)  (h - 2, 1)  (h - 2, 2)  (h - 2, 3)
+//                             .................................
+// level 0            (0, 0)  (0, 1)  (0, 2)  (0, 3) ... (0, numberTransactions - 1)
+
+
+
+// We travel the tree in depth-first order. A vBits[i] is true if node i-th in DFS
+// is parent of leaf node which is we want to verify, otherwise this value is false.
+// vHash store hash value at node i in DFS order. Follow the vBits and vHash,
+// we can rebuild the tree and extract merkle path we want.
+type partialMerkleTreeData struct {
+	numberTransactions uint 
+	vBits              []bool
+	vHash              []*chainhash.Hash
+}
+
+// Merkle proof use for rebuild the merkle tree and extract merkle proof for
+// single transaction
+type MerkleProof struct {
+	// merkle node value at this sub tree
+	merkleRoot chainhash.Hash
+	// merkle path if the txID we want to build merkle proof in this subtree
+	// if not this is empty
+	merklePath []chainhash.Hash
+	// transaction index. We use this for check "left, right" when
+	// compute merkle root.
```

Comment:
please explain how the left/right direction is encoded. Is 0 bit left? Which bit is leaf (most significant or the least significant)?

Reply:
We need details. IIRC, `transactionIndex >> h % 2 == 0` if at height `h` w our node is on left, right, correct?

Reply:
Yes, I add more comments about this.